include/drivers/rtio.h

```diff
@@ -0,0 +1,487 @@
+/*
+ * Copyright (c) 2019  Thomas Burdick <thomas.burdick@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_RTIO_H_
+#define ZEPHYR_INCLUDE_DRIVERS_RTIO_H_
+/**
+ * @brief RTIO Interface
+ * @defgroup rtio_interface RTIO Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <kernel.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @file
+ * @brief Real-Time IO device API for moving bytes around fast
+ *
+ * Each Real-Time IO device provides a way to either read, write, or both
+ * blocks of memory allocated from a static pool by the application. The layout
+ * of the each block is understood implicitly by the author. Each device may
+ * provide functionality to safely encode or decode blocks of memory.
+ */
+
+/**
+ * @brief Block of memory
+ *
+ * This block of memory has important metadata commonly useful such as
+ * byte layout, and timestamps of when a read or write began and ended.
+ */
+struct rtio_block {
+	/** Timestamp in cycles from k_cycle_get_32() marking when a read or
+	 * write began
+	 */
+	u32_t begin_tstamp;
+
+	/** Timestamp in cycles from k_cycle_get_32() marking when a read or
+	 * write was complete
+	 */
+	u32_t end_tstamp;
+
+	/** Byte layout designator given by each driver to specify
+	 * the layout of the contained buffer
+	 */
+	u32_t layout;
+
+	/** Length of data in this buffer. */
+	u16_t len;
+
+	/** Amount of data this buffer can store. */
+	u16_t size;
+
```

Comment:
Could we have len and size as u32_t?

Reply:
Using size_t for length/size of the buffer would probably be appropriate here.

Reply:
We most certainly can do size_t's, and I think it makes sense

Reply:
I've switched everything to size_t to start. Self reviewing this I'm not sure I like that the policy describes size in bytes rather than sample sets or values. Especially given my additional work to provide a sensor.h like API for reading data.

Reply:
Offhand I think it makes sense and at this layer is most flexible to use bytes.  Any other measure would complicate/prevent using it for streams of variable-sized observations.  I have some experience with streams of sampled data passing through a transformation system: pretty soon you want to add sample-aligned metadata, and then fork and join nodes that require processing streams of events where each packet carries its own timestamp and internal metadata.

Reply:
>> I'm not sure I like that the policy describes size in bytes rather than sample sets or values

> Offhand I think it makes sense and at this layer is most flexible to use bytes.

I agree with @pabigot (and your first instinct, @bfrog) on this one. You can always add sensor-specific macros to divide bytes to number of samples.

Reply:
Perfect, will leave it be then